utils/io-wrappers.cc

```diff
@@ -0,0 +1,245 @@
+            size_t res = 0;
+            for (auto& iv : iov) {
+                res += co_await do_write_dma(pos + res, iv.iov_base, iv.iov_len);
+            }
+            co_return res;
+        }
+        future<> truncate(uint64_t length) override {
+            if (length < _pos) {
+                // bufpos should be a full buffer here.
+                if (length < (_pos - _bufpos)) {
+                    throw std::invalid_argument("Cannot truncate below one disk page of current position");
+                }
+                _bufpos -= (_pos - length);
+            }
+            if (_bufpos != 0) {
+                co_await put(_buffer, _bufpos);
+                _bufpos = 0;
+            }
+            // if we truncate upwards, just write zeroes.
+            uint64_t bp = 0;
+            while (length > _pos) {
+                auto rem = std::min(length - _pos, sizeof(_buffer));
+                if (rem > bp) {
+                    std::fill(_buffer, _buffer + rem , 0);
+                    bp = rem;
+                }
+                co_await put(_buffer, rem);
+                _pos += rem;
+            }
+        }
+        future<> close() override {
+            co_await truncate(_pos);
+            co_await _sink.flush();
+            co_await _sink.close();
+        }
+    };
+    return file{make_shared<data_sink_file_impl>(std::move(sink))};
+}
+
+file create_noop_file() {
+    return file{make_shared<noop_file_impl>()};
+}
+
+data_sink create_memory_sink(std::vector<seastar::temporary_buffer<char>>& bufs) {
+    // TODO: move to seastar. Based on memory_data_sink, but allowing us
+    // to actually move away the buffers later. I don't want to modify
+    // util classes in an enterprise patch.
+    class buffer_data_sink_impl : public data_sink_impl {
+        std::vector<temporary_buffer<char>>& _bufs;
+    public:
+        buffer_data_sink_impl(std::vector<temporary_buffer<char>>& bufs)
+            : _bufs(bufs)
+        {}
+        future<> put(net::packet p) override {
+            return put(p.release());
+        }
+        future<> put(std::vector<temporary_buffer<char>> bufs) override {
+            for (auto&& buf : bufs) {
+                _bufs.emplace_back(std::move(buf));
+            }
+            return make_ready_future<>();
+        }
+        future<> put(temporary_buffer<char> buf) override {
+            _bufs.emplace_back(std::move(buf));
+            return make_ready_future<>();
+        }
+        future<> flush() override {
+            return make_ready_future<>();
+        }
+        future<> close() override {
+            return make_ready_future<>();
+        }
+        size_t buffer_size() const noexcept override {
+            return 128*1024;
+        }
+    };
+    return data_sink(std::make_unique<buffer_data_sink_impl>(bufs));
+}
+
+data_source create_memory_source(std::vector<seastar::temporary_buffer<char>> bufs) {
```

Comment:
This seems to duplicate the `create_stream_sources`'s `sstable_stream_source_impl::buffer_data_source_impl`

Reply:
It does. And once this PR is commited, I will remove them from sstable and use these shared versions. But I prefer to mix the matters minimally.